UM/Settings/MachineManager.py

```diff
@@ -150,7 +200,42 @@ def setActiveMachineInstance(self, machine):
 
         self._updateSettingVisibility(setting_visibility)
 
+        self._protect_working_profile = True
+        profile = self.findProfile(machine.getActiveProfileName(), instance = self._active_machine)
+        if profile:
+            self.setActiveProfile(profile)
+        else:
+            for profile in self._profiles:
+                self.setActiveProfile(profile) #default to first profile you can find
+                break
+
+        if self._active_machine.hasMaterials():
+            material = self._active_machine.getMaterialName()
+            available_materials = self.getAllMachineMaterials(self._active_machine.getName())
+            if not material or (len(available_materials) > 0 and material not in available_materials):
+                if "PLA" in available_materials:
+                    self._active_machine.setMaterialName("PLA")
+                else:
+                    self._active_machine.setMaterialName(available_materials[0])
+
         self.activeMachineInstanceChanged.emit()
+        self._protect_working_profile = False
+
+    def setActiveMaterial(self, material):
+        if not self._active_machine:
+            return
+
+        emit = False
+        if material != self._active_machine.getMaterialName():
+            self._active_machine.setMaterialName(material)
+            emit = True
+
+        material_profile = self.findProfile(material, type = "material", instance = self._active_machine)
+        if material_profile:
+            self._active_machine.getWorkingProfile().mergeSettingsFrom(material_profile, reset = False)
```

Comment:
> Since it changes the active machine.
It doesn't. It changes the working profile. The "material_profile" here is a partial profile, that gets merged into the current settings in the working profile. 

Needs more documentation.